File: apps/workflow-a/src/handlers/validate.handler.ts
import type { Context } from 'aws-lambda';
import { createLogger, withSpan } from '@org/shared-observability';
import { validateProposta } from '@org/shared-validation';
import { createRuleEngine, minimumValueRule } from '@org/shared-rules-engine';
import { checkFraud } from '@org/shared-fraud-engine';
import {
  MIN_VALOR,
  WORKFLOW_NAME,
  type PropostaEvent,
  type StepResult,
} from './types.js';

const logger = createLogger({ service: WORKFLOW_NAME, handler: 'validate' });

const ruleEngine = createRuleEngine().add(minimumValueRule(MIN_VALOR));

/**
 * Validates the proposal, runs the shared rule + fraud engines and decides the
 * next status:
 *   - invalid data or failing rules -> REJECTED
 *   - high fraud risk               -> IN_REVIEW (manual review)
 *   - otherwise                     -> APPROVED
 */
export const handler = async (
  event: PropostaEvent,
  context: Context
): Promise<StepResult> =>
  withSpan(
    'validate',
    async () => {
      const proposta = event.payload;

      const validation = validateProposta(proposta);
      const rules = await ruleEngine.run(proposta);
      const fraud = checkFraud({
        documento: proposta.cliente.documento,
        valor: proposta.valor,
      });

      let status: StepResult['status'];
      if (!validation.valid || !rules.passed) {
        status = 'REJECTED';
      } else if (fraud.risk === 'high') {
        status = 'IN_REVIEW';
      } else {
        status = 'APPROVED';
      }

      logger.info('proposta.validated', {
        requestId: context.awsRequestId,
        propostaId: proposta.id,
        status,
        fraudScore: fraud.score,
        ruleFailures: rules.failures,
      });

      return {
        workflow: WORKFLOW_NAME,
        step: 'validate',
        propostaId: proposta.id,
        status,
        detail: {
          validationErrors: validation.errors,
          ruleFailures: rules.failures,
          fraud,
        },
      };
    },
    logger
  );

File: apps/workflow-b/Handlers/ValidateHandler.cs
using Amazon.Lambda.Core;
using Acme.Shared.Domain;
using Acme.Shared.FraudEngine;
using Acme.Shared.Observability;
using Acme.Shared.RulesEngine;
using Acme.Shared.Validation;

namespace Acme.WorkflowB.Handlers;

/// <summary>
/// Validates the proposal, runs the shared rule + fraud engines and decides the
/// next status:
///   - invalid data or failing rules -> Rejected
///   - high fraud risk               -> InReview (manual review)
///   - otherwise                     -> Approved
/// </summary>
public sealed class ValidateHandler
{
    private static readonly IStructuredLogger Logger = StructuredLogger.Create(
        new Dictionary<string, object?> { ["service"] = WorkflowConstants.Name, ["handler"] = "validate" });

    private static readonly RuleEngine<Proposta> Rules =
        new RuleEngine<Proposta>().Add(new MinimumValueRule(WorkflowConstants.MinValor));

    public Task<StepResult> Handle(WorkflowEvent<Proposta> evt, ILambdaContext context) =>
        Tracing.WithSpanAsync("validate", Logger, span =>
        {
            var proposta = evt.Payload;

            var validation = Validators.ValidateProposta(proposta);
            var rules = Rules.Run(proposta);
            var fraud = FraudChecker.CheckFraud(proposta.Cliente.Documento, proposta.Valor);

            var status = (validation.Valid && rules.Passed, fraud.Risk) switch
            {
                (false, _) => Status.Rejected,
                (true, RiskLevel.High) => Status.InReview,
                _ => Status.Approved,
            };

            span.Info("proposta.validated", new Dictionary<string, object?>
            {
                ["requestId"] = context.AwsRequestId,
                ["propostaId"] = proposta.Id,
                ["status"] = status.ToString(),
                ["fraudScore"] = fraud.Score,
                ["ruleFailures"] = rules.Failures,
            });

            var detail = new Dictionary<string, object?>
            {
                ["validationErrors"] = validation.Errors,
                ["ruleFailures"] = rules.Failures,
                ["fraudScore"] = fraud.Score,
                ["fraudRisk"] = fraud.Risk.ToString(),
            };
            return Task.FromResult(new StepResult(
                WorkflowConstants.Name, "validate", proposta.Id, status, detail));
        });
}
